Relax GOT indirection into PC-relative addressing (#397)
riscv-non-isa/riscv-elf-psabi-doc: riscv-elf.adoc

  `R_RISCV_PCREL_LO12_I` points.

  - The symbol pointed to by `R_RISCV_PCREL_LO12_I` is at the location to
  which `R_RISCV_GOT_HI20` refers.

  - The symbol's address is a link-time constant.

  - If the symbol is absolute, its address is within `0x0` ~ `0x7ff`.
  Note that an unresolved weak symbol satisfies this condition because
  such symbol is handled as if it were an absolute symbol at address 0.

  - If the symbol is relative, the offset between the location to
  which `R_RISCV_GOT_HI20` refers and the target symbol is within +-2GiB.

  Relaxation::
  - Instruction sequence associated with `R_RISCV_GOT_HI20` and
  `R_RISCV_PCREL_LO12_I` can be rewritten to an `c.li`, `addi` or
  `auipc` + `addi` to materialize the symbol's address directly in a
  register, eliminating a load from the GOT.

  - If this relaxation eliminates all references to the symbol's GOT slot,
  the linker may opt not to create a GOT slot for that symbol.

  Example::
+
--
Relaxation candidate:
[,asm]
----
label:
    auipc   t0, 0      # R_RISCV_GOT_HI20 (symbol), R_RISCV_RELAX
    ld      t0, 0(t0)  # R_RISCV_PCREL_LO12_I (label), R_RISCV_RELAX
----

> Relaxation result (absolute symbol whose address is within `0x0` ~ `0x1f`

Reviewer: > whose address is within `0x0` ~ `0x1f`

Where does this come from?

Author: Sorry, `c.lui` was a typo of `c.addi`.

Reviewer: Why is the address 0x0 to 0x1f?

Author: Because `c.li` can set only a signed 6 bit value.

Reviewer 2: more precisely range: "0x0 ~ 0x1f or 0xffff ffff ffff ff20 ~ 0xffffffffffffffff for RV64 and 0xffffff20 ~ 0xffffffff for RV32." since it's sign-extended value?

NOTE: Tag_RISCV_x3_reg_usage is treated as 0 if it is not present.

==== GOT load relaxation

  Target Relocation:: R_RISCV_GOT_HI20, R_RISCV_PCREL_LO12_I

  Description:: This relaxation can relax a GOT indirection into load
  immediate or PC-relative addressing. This relaxation is intended to
  optimize the `lga` assembly pseudo-instruction (and thus `la` for
  PIC objects), which loads a symbol's address from a GOT entry with
  an `auipc` + `l[w|d]` instruction pair.

  Condition::
  - Both `R_RISCV_GOT_HI20` and `R_RISCV_PCREL_LO12_I` are marked with
  `R_RISCV_RELAX`.

  - The symbol pointed to by `R_RISCV_PCREL_LO12_I` is at the location to
  which `R_RISCV_GOT_HI20` refers.

>   - If the symbol is absolute, its address is within `0x0` ~ `0x7ff`.

Reviewer 2: more precisely range: "`0x0` ~ `0x7ff` or `0xfffffffffffff800` ~ `0xffffffffffffffff` for RV64 and `0xfffff800` ~ `0xffffffff` for RV32." since it's sign-extended value.

Author: This is mainly for undefined weak symbols whose address is zero, but yeah, we can technically use it for the highest addresses. I'll update the proposal.

NOTE: Tag_RISCV_x3_reg_usage is treated as 0 if it is not present.

==== GOT load relaxation

  Target Relocation:: R_RISCV_GOT_HI20, R_RISCV_PCREL_LO12_I

  Description:: This relaxation can relax a GOT indirection into load
  immediate or PC-relative addressing. This relaxation is intended to

>   optimize the `la`/`lga` assembly pseudo-instruction, which loads a

Reviewer 3: ```suggestion
  optimize the `lga` assembly pseudo-instruction (and thus `la` for PIC objects), which loads a
```

>   symbol's address from a GOT entry with an `auipc` + `ld` instruction

Reviewer 3: ```suggestion
  symbol's address from a GOT entry with an `auipc` + `l[w|d]` instruction
```

  pair.

>   Condition::

Reviewer 3: The language in here needs tweaking to more obviously cover the case when a single GOT_HI20 is shared by multiple PCREL_LO12_I

Author: I don't think we want to handle such case. This optimization is to relax `la` and not others. It looks like a GOT_HI20 is almost always paired with a PCREL_LO_I that immediately follows, so this should be sufficient.

Reviewer 3: LLVM does not model them as a pair, it models them as independent instructions. In practice you will put them together to benefit from macro-op fusion, but there's no requirement to do so in the ABI or in LLVM's internals, so the ABI should not start to introduce arbitrary restrictions that have no technical grounding other than being marginally more annoying to write a specification for. Please specify the actual necessary restrictions.

  - Both `R_RISCV_GOT_HI20` and `R_RISCV_PCREL_LO12_I` are marked
  with `R_RISCV_RELAX`.

  - `R_RISCV_GOT_HI20` refers to the location 4 bytes before where

>   `R_RISCV_PCREL_LO12_I` points.

Reviewer 3: Why? Who cares where they are relative to each other.

Author: I wanted to exclude the pattern that one GOT_HI20 is shared by multiple PCREL_LO_I because it'd complicates the specification a lot. This optimization focuses on optimizing `la` pseudo instruction

Reviewer 3: It's not complicated at all, just require that the GOT_HI20 and all referencing PCREL_LO12_I's are relaxable

Author: Done

  - The symbol pointed to by `R_RISCV_PCREL_LO12_I` is at the location to
  which `R_RISCV_GOT_HI20` refers.

>   - The symbol's address is a link-time constant.

Reviewer 3: That's not what you mean. What you mean is that it's bound at link time. Otherwise a PIE, or DSO with suitable binding, wouldn't be able to use it, as the object would have an unknown load base and thus the address wouldn't be a link-time constant, just its PC-relative offset.

Author: I meant to say "The symbol's PC-relative address is a link-time constant." Does this make sense to you?

Reviewer 3: That's not true for absolute symbols. Just do it in terms of "the symbol is bound at link time to within the object / is defined in the object and not preemptible" or similar, because that's what the important part is.

Author: The actual condition is a bit more complicated due to GNU ifunc. Let me rearrange this section to explain it more clearly.

  - If the symbol is absolute, its address is within `0x0` ~ `0x7ff`.

>   Note that an unresolved weak symbol satisfies this condition because

Reviewer 3: ```suggestion
  Note that an undefined weak symbol satisfies this condition because
```

>   such symbol is handled as if it were an absolute symbol at address 0.

Reviewer 3: ```suggestion
  such a symbol is handled as if it were an absolute symbol at address 0.
```

  - If the symbol is relative, it's bound at link time to be within the
  object. It should not be of the GNU ifunc type. Additionally, the offset
  between the location to which `R_RISCV_GOT_HI20` refers and the target
  symbol should be within a range of +-2GiB.

  Relaxation::
  - The `auipc` instruction associated with `R_RISCV_GOT_HI20` can be
  removed if the symbol is absolute.

  - The instruction or instructions associated with `R_RISCV_PCREL_LO12_I`
  can be rewritten to either `c.li` or `addi` to materialize the symbol's
  address directly in a register.

  - If this relaxation eliminates all references to the symbol's GOT slot,
  the linker may opt not to create a GOT slot for that symbol.

  Example::
+
--
Relaxation candidate:
[,asm]
----
label:
    auipc   tX, 0      # R_RISCV_GOT_HI20 (symbol), R_RISCV_RELAX
    l[w|d]  tY, 0(tX)  # R_RISCV_PCREL_LO12_I (label), R_RISCV_RELAX
----

Relaxation result (absolute symbol whose address can be represented as
a 6-bit signed integer and if the RVC instruction is permitted):

[,asm]
----
    c.li    tY, <symbol-value>
----

Relaxation result (absolute symbol whose address cannot be represented

> as a 6-bit signed integer or if the RVC instruction is not permitted):

Reviewer: This is 12-bit: [-2048, 2047]

Author: Fixed. Thanks!

Note that the use case of this relaxation pattern is to relax a GOT-indirect reference to an unresolved weak symbol in an environment that the C extension is not available.

Author: This resulted in a wrong condition, so I rewrote the sentence.
